delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/files/batch/BatchCopy.scala

```diff
@@ -0,0 +1,125 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.storage.files.batch
+
+import cats.data.NonEmptyList
+import cats.effect.IO
+import cats.implicits.toFunctorOps
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.UUIDF
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.FetchFileResource
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.model.FileRejection._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.model._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.routes.CopyFileSource
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.Storage.{DiskStorage, RemoteDiskStorage}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.{Storage, StorageType}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.StorageFileRejection.CopyFileRejection
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.StorageFileRejection.CopyFileRejection._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.disk.{DiskCopyDetails, DiskStorageCopyFiles}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.remote.RemoteDiskStorageCopyFiles
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.remote.client.model.RemoteDiskCopyDetails
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.{FetchStorage, StoragesStatistics}
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.sdk.acls.AclCheck
+import ch.epfl.bluebrain.nexus.delta.sdk.error.ServiceError.AuthorizationFailed
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.Caller
+import shapeless.syntax.typeable.typeableOps
+
+trait BatchCopy {
+  def copyFiles(source: CopyFileSource, destStorage: Storage)(implicit
+      c: Caller
+  ): IO[NonEmptyList[FileAttributes]]
+}
+
+object BatchCopy {
+  def mk(
+      fetchFile: FetchFileResource,
+      fetchStorage: FetchStorage,
+      aclCheck: AclCheck,
+      storagesStatistics: StoragesStatistics,
+      diskCopy: DiskStorageCopyFiles,
+      remoteDiskCopy: RemoteDiskStorageCopyFiles
+  )(implicit uuidF: UUIDF): BatchCopy = new BatchCopy {
+
+    override def copyFiles(source: CopyFileSource, destStorage: Storage)(implicit
+        c: Caller
+    ): IO[NonEmptyList[FileAttributes]] =
+      destStorage match {
+        case disk: Storage.DiskStorage         => copyToDiskStorage(source, disk)
```

Comment:
Maybe better as 2 classes dealing with each case alone

Reply:
Thought about this, it felt like just more verbosity since most of the logic is shared. Also do not want to change the tests again 👀  (they're maybe probably a little overkill tbh)